File: armadilha_cdi/services/calculations.py
from __future__ import annotations

from bisect import bisect_right
from datetime import date

from armadilha_cdi.config import (
    EARLIEST_SUPPORTED_DATE,
    MAX_MARKET_DATE_FALLBACK_DAYS,
    MAX_USD_FALLBACK_DAYS,
)
from armadilha_cdi.exceptions import DataUnavailableError, DomainValidationError
from armadilha_cdi.models import CalculationResult, QuoteLookup


def validate_inputs(start_date: date, end_date: date, initial_brl: float) -> None:
    if start_date < EARLIEST_SUPPORTED_DATE:
        raise DomainValidationError(
            "A data inicial deve ser em ou posterior a "
            f"{EARLIEST_SUPPORTED_DATE.strftime('%d/%m/%Y')}, "
            "quando o real brasileiro entrou em circulacao."
        )
    if end_date <= start_date:
        raise DomainValidationError("A data final deve ser maior que a data inicial.")
    if initial_brl <= 0:
        raise DomainValidationError("O valor inicial deve ser maior que zero.")


class QuoteResolver:
    """Resolve USD/BRL quotes using the nearest previous available date."""

    def __init__(
        self,
        usd_rates: dict[str, float],
        max_days_back: int = MAX_USD_FALLBACK_DAYS,
        min_date: date | None = None,
    ) -> None:
        self.max_days_back = max_days_back
        self.min_date = min_date
        self._values_by_date: dict[date, float] = {}

        for iso_date, value in usd_rates.items():
            try:
                parsed_date = date.fromisoformat(iso_date)
                if self.min_date is not None and parsed_date < self.min_date:
                    continue
                self._values_by_date[parsed_date] = float(value)
            except (TypeError, ValueError):
                continue

        self._ordered_dates = sorted(self._values_by_date)

    def lookup(self, target_date: date) -> QuoteLookup:
        index = bisect_right(self._ordered_dates, target_date) - 1
        if index >= 0:
            effective_date = self._ordered_dates[index]
            if (target_date - effective_date).days <= self.max_days_back:
                return QuoteLookup(
                    requested_date=target_date,
                    effective_date=effective_date,
                    value=self._values_by_date[effective_date],
                )

        raise DataUnavailableError(
            "Nao foi encontrada cotacao USD/BRL suficiente para o periodo selecionado."
        )


class MarketDateResolver:
    """Resolve official market dates using the nearest previous available date."""

    def __init__(
        self,
        series: dict[str, float],
        label: str,
        max_days_back: int = MAX_MARKET_DATE_FALLBACK_DAYS,
        min_date: date | None = None,
    ) -> None:
        self.label = label
        self.max_days_back = max_days_back
        self.min_date = min_date
        self._ordered_dates: list[date] = []

        for iso_date in series:
            try:
                parsed_date = date.fromisoformat(str(iso_date))
                if self.min_date is not None and parsed_date < self.min_date:
                    continue
                self._ordered_dates.append(parsed_date)
            except (TypeError, ValueError):
                continue

        self._ordered_dates = sorted(set(self._ordered_dates))

    def lookup(
        self,
        target_date: date,
        allow_forward_if_before_first: bool = False,
    ) -> date:
        index = bisect_right(self._ordered_dates, target_date) - 1
        if index >= 0:
            effective_date = self._ordered_dates[index]
            if (target_date - effective_date).days <= self.max_days_back:
                return effective_date

        if allow_forward_if_before_first and self._ordered_dates:
            effective_date = self._ordered_dates[0]
            if (
                target_date <= effective_date
                and (effective_date - target_date).days <= self.max_days_back
            ):
                return effective_date

        raise DataUnavailableError(
            f"Nao foi encontrado dado de {self.label} suficiente para o periodo selecionado."
        )


def lookup_quote_with_fallback(
    usd_rates: dict[str, float],
    target_date: date,
    max_days_back: int = MAX_USD_FALLBACK_DAYS,
) -> QuoteLookup:
    return QuoteResolver(usd_rates=usd_rates, max_days_back=max_days_back).lookup(target_date)


def resolve_cdi_period(
    cdi_rates: dict[str, float],
    start_date: date,
    end_date: date,
) -> tuple[date, date]:
    resolver = MarketDateResolver(
        series=cdi_rates,
        label="CDI",
        min_date=EARLIEST_SUPPORTED_DATE,
    )
    effective_start_date = resolver.lookup(
        start_date,
        allow_forward_if_before_first=True,
    )
    effective_end_date = resolver.lookup(end_date)

    if effective_end_date <= effective_start_date:
        raise DataUnavailableError("Nao ha dias uteis de CDI suficientes para o periodo informado.")

    return effective_start_date, effective_end_date


def calculate_cdi_factor(
    cdi_rates: dict[str, float],
    start_date: date,
    end_date: date,
) -> tuple[float, int]:
    factor = 1.0
    start_key = start_date.isoformat()
    end_key = end_date.isoformat()

    window_rates: list[tuple[str, float]] = []
    for iso_date, rate in cdi_rates.items():
        if not isinstance(iso_date, str):
            continue
        if start_key <= iso_date < end_key:
            try:
                date.fromisoformat(iso_date)
                window_rates.append((iso_date, float(rate)))
            except (TypeError, ValueError):
                continue

    for _, rate in sorted(window_rates):
        factor *= 1 + (rate / 100)

    days_used = len(window_rates)

    if days_used == 0:
        raise DataUnavailableError("Nao ha dados de CDI suficientes para o periodo informado.")

    return factor, days_used


def calculate_result(
    start_date: date,
    end_date: date,
    initial_brl: float,
    cdi_rates: dict[str, float],
    usd_rates: dict[str, float],
) -> CalculationResult:
    validate_inputs(start_date=start_date, end_date=end_date, initial_brl=initial_brl)
    effective_start_date, effective_end_date = resolve_cdi_period(
        cdi_rates=cdi_rates,
        start_date=start_date,
        end_date=end_date,
    )

    cdi_factor, cdi_days_used = calculate_cdi_factor(
        cdi_rates=cdi_rates,
        start_date=effective_start_date,
        end_date=effective_end_date,
    )
    quote_resolver = QuoteResolver(usd_rates=usd_rates, min_date=EARLIEST_SUPPORTED_DATE)
    initial_quote = quote_resolver.lookup(effective_start_date)
    final_quote = quote_resolver.lookup(effective_end_date)

    final_brl = initial_brl * cdi_factor
    initial_usd = initial_brl / initial_quote.value
    final_usd_with_cdi = final_brl / final_quote.value
    real_usd_return_percentage = ((final_usd_with_cdi / initial_usd) - 1) * 100

    return CalculationResult(
        start_date=start_date,
        end_date=end_date,
        effective_start_date=effective_start_date,
        effective_end_date=effective_end_date,
        initial_brl=initial_brl,
        final_brl=final_brl,
        cdi_factor=cdi_factor,
        cdi_percentage=(cdi_factor - 1) * 100,
        initial_usd=initial_usd,
        final_usd_with_cdi=final_usd_with_cdi,
        initial_usdbrl=initial_quote.value,
        final_usdbrl=final_quote.value,
        initial_fx_date=initial_quote.effective_date,
        final_fx_date=final_quote.effective_date,
        real_usd_return_percentage=real_usd_return_percentage,
        cdi_days_used=cdi_days_used,
    )

File: public/assets/calculations.js
export const EARLIEST_SUPPORTED_DATE = "1994-07-01";
export const MAX_USD_FALLBACK_DAYS = 15;
export const MAX_MARKET_DATE_FALLBACK_DAYS = 15;
export const BUSINESS_DAYS_PER_YEAR = 252;
export const BUSINESS_DAYS_PER_MONTH = 22;

const ISO_DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/;
const MS_PER_DAY = 24 * 60 * 60 * 1000;

export class DomainValidationError extends Error {
  constructor(message) {
    super(message);
    this.name = "DomainValidationError";
  }
}

export class DataUnavailableError extends Error {
  constructor(message) {
    super(message);
    this.name = "DataUnavailableError";
  }
}

export function isIsoDate(value) {
  if (typeof value !== "string" || !ISO_DATE_PATTERN.test(value)) {
    return false;
  }
  const parsed = new Date(`${value}T00:00:00Z`);
  return !Number.isNaN(parsed.getTime()) && parsed.toISOString().slice(0, 10) === value;
}

function parseIsoDate(value) {
  if (!isIsoDate(value)) {
    throw new DomainValidationError("Use datas no formato YYYY-MM-DD.");
  }
  return new Date(`${value}T00:00:00Z`);
}

function daysBetween(startDate, endDate) {
  return Math.round((parseIsoDate(endDate).getTime() - parseIsoDate(startDate).getTime()) / MS_PER_DAY);
}

function normalizeSeries(series, minDate = null) {
  const valuesByDate = new Map();
  if (!series || typeof series !== "object") {
    return valuesByDate;
  }

  for (const [rawDate, rawValue] of Object.entries(series)) {
    if (!isIsoDate(rawDate)) {
      continue;
    }
    if (minDate !== null && rawDate < minDate) {
      continue;
    }
    const value = Number(rawValue);
    if (!Number.isFinite(value)) {
      continue;
    }
    valuesByDate.set(rawDate, value);
  }
  return valuesByDate;
}

function sortedDatesFromMap(valuesByDate) {
  return Array.from(valuesByDate.keys()).sort();
}

function bisectRight(sortedValues, target) {
  let left = 0;
  let right = sortedValues.length;
  while (left < right) {
    const middle = Math.floor((left + right) / 2);
    if (sortedValues[middle] <= target) {
      left = middle + 1;
    } else {
      right = middle;
    }
  }
  return left;
}

export function validateInputs(startDate, endDate, initialBrl) {
  if (!isIsoDate(startDate) || !isIsoDate(endDate)) {
    throw new DomainValidationError("Use datas no formato YYYY-MM-DD.");
  }
  if (startDate < EARLIEST_SUPPORTED_DATE) {
    throw new DomainValidationError(
      "A data inicial deve ser em ou posterior a 01/07/1994, quando o real brasileiro entrou em circulação."
    );
  }
  if (endDate <= startDate) {
    throw new DomainValidationError("A data final deve ser maior que a data inicial.");
  }
  if (!(Number(initialBrl) > 0)) {
    throw new DomainValidationError("O valor inicial deve ser maior que zero.");
  }
}

export class QuoteResolver {
  constructor(usdRates, maxDaysBack = MAX_USD_FALLBACK_DAYS, minDate = null) {
    this.maxDaysBack = maxDaysBack;
    this.valuesByDate = normalizeSeries(usdRates, minDate);
    this.orderedDates = sortedDatesFromMap(this.valuesByDate);
  }

  lookup(targetDate) {
    if (!isIsoDate(targetDate)) {
      throw new DomainValidationError("Use datas no formato YYYY-MM-DD.");
    }
    const index = bisectRight(this.orderedDates, targetDate) - 1;
    if (index >= 0) {
      const effectiveDate = this.orderedDates[index];
      if (daysBetween(effectiveDate, targetDate) <= this.maxDaysBack) {
        return {
          requested_date: targetDate,
          effective_date: effectiveDate,
          value: this.valuesByDate.get(effectiveDate)
        };
      }
    }
    throw new DataUnavailableError(
      "Não foi encontrada cotação USD/BRL suficiente para o período selecionado."
    );
  }
}

export class MarketDateResolver {
  constructor(series, label, maxDaysBack = MAX_MARKET_DATE_FALLBACK_DAYS, minDate = null) {
    this.label = label;
    this.maxDaysBack = maxDaysBack;
    this.orderedDates = sortedDatesFromMap(normalizeSeries(series, minDate));
  }

  lookup(targetDate, allowForwardIfBeforeFirst = false) {
    if (!isIsoDate(targetDate)) {
      throw new DomainValidationError("Use datas no formato YYYY-MM-DD.");
    }
    const index = bisectRight(this.orderedDates, targetDate) - 1;
    if (index >= 0) {
      const effectiveDate = this.orderedDates[index];
      if (daysBetween(effectiveDate, targetDate) <= this.maxDaysBack) {
        return effectiveDate;
      }
    }

    if (allowForwardIfBeforeFirst && this.orderedDates.length > 0) {
      const effectiveDate = this.orderedDates[0];
      if (targetDate <= effectiveDate && daysBetween(targetDate, effectiveDate) <= this.maxDaysBack) {
        return effectiveDate;
      }
    }

    throw new DataUnavailableError(
      `Não foi encontrado dado de ${this.label} suficiente para o período selecionado.`
    );
  }
}

export function lookupQuoteWithFallback(usdRates, targetDate, maxDaysBack = MAX_USD_FALLBACK_DAYS) {
  return new QuoteResolver(usdRates, maxDaysBack).lookup(targetDate);
}

export function resolveCdiPeriod(cdiRates, startDate, endDate) {
  const resolver = new MarketDateResolver(
    cdiRates,
    "CDI",
    MAX_MARKET_DATE_FALLBACK_DAYS,
    EARLIEST_SUPPORTED_DATE
  );
  const effectiveStartDate = resolver.lookup(startDate, true);
  const effectiveEndDate = resolver.lookup(endDate);

  if (effectiveEndDate <= effectiveStartDate) {
    throw new DataUnavailableError("Não há dias úteis de CDI suficientes para o período informado.");
  }

  return [effectiveStartDate, effectiveEndDate];
}

export function calculateCdiFactor(cdiRates, startDate, endDate) {
  const normalized = normalizeSeries(cdiRates);
  const windowRates = [];
  for (const [isoDate, rate] of normalized.entries()) {
    if (startDate <= isoDate && isoDate < endDate) {
      windowRates.push([isoDate, rate]);
    }
  }
  windowRates.sort((left, right) => left[0].localeCompare(right[0]));

  let factor = 1.0;
  for (const [, rate] of windowRates) {
    factor *= 1 + rate / 100;
  }

  if (windowRates.length === 0) {
    throw new DataUnavailableError("Não há dados de CDI suficientes para o período informado.");
  }
  return [factor, windowRates.length];
}

export function calculateEquivalentRatePercentage(
  periodPercentage,
  periodBusinessDays,
  equivalentBusinessDays
) {
  if (periodBusinessDays <= 0) {
    throw new Error("period_business_days must be greater than zero.");
  }
  if (equivalentBusinessDays <= 0) {
    throw new Error("equivalent_business_days must be greater than zero.");
  }
  const periodFactor = 1 + periodPercentage / 100;
  const equivalentFactor = periodFactor ** (equivalentBusinessDays / periodBusinessDays);
  return (equivalentFactor - 1) * 100;
}

export function calculateResult({ startDate, endDate, initialBrl, cdiRates, usdRates }) {
  const normalizedInitialBrl = Number(initialBrl);
  validateInputs(startDate, endDate, normalizedInitialBrl);
  const [effectiveStartDate, effectiveEndDate] = resolveCdiPeriod(cdiRates, startDate, endDate);
  const [cdiFactor, cdiDaysUsed] = calculateCdiFactor(cdiRates, effectiveStartDate, effectiveEndDate);

  const quoteResolver = new QuoteResolver(usdRates, MAX_USD_FALLBACK_DAYS, EARLIEST_SUPPORTED_DATE);
  const initialQuote = quoteResolver.lookup(effectiveStartDate);
  const finalQuote = quoteResolver.lookup(effectiveEndDate);

  const finalBrl = normalizedInitialBrl * cdiFactor;
  const initialUsd = normalizedInitialBrl / initialQuote.value;
  const finalUsdWithCdi = finalBrl / finalQuote.value;
  const realUsdReturnPercentage = (finalUsdWithCdi / initialUsd - 1) * 100;

  return {
    start_date: startDate,
    end_date: endDate,
    effective_start_date: effectiveStartDate,
    effective_end_date: effectiveEndDate,
    initial_brl: normalizedInitialBrl,
    final_brl: finalBrl,
    cdi_factor: cdiFactor,
    cdi_percentage: (cdiFactor - 1) * 100,
    initial_usd: initialUsd,
    final_usd_with_cdi: finalUsdWithCdi,
    initial_usdbrl: initialQuote.value,
    final_usdbrl: finalQuote.value,
    initial_fx_date: initialQuote.effective_date,
    final_fx_date: finalQuote.effective_date,
    real_usd_return_percentage: realUsdReturnPercentage,
    cdi_days_used: cdiDaysUsed
  };
}

export function buildChartSeries({ startDate, endDate, cdiRates, usdRates, initialBrl }) {
  const normalizedInitialBrl = Number(initialBrl);
  validateInputs(startDate, endDate, normalizedInitialBrl);
  const [effectiveStartDate, effectiveEndDate] = resolveCdiPeriod(cdiRates, startDate, endDate);
  const quoteResolver = new QuoteResolver(usdRates, MAX_USD_FALLBACK_DAYS, EARLIEST_SUPPORTED_DATE);
  const initialQuote = quoteResolver.lookup(effectiveStartDate);
  const cdiValues = normalizeSeries(cdiRates);

  const timeline = sortedDatesFromMap(cdiValues).filter(
    (currentDate) => effectiveStartDate <= currentDate && currentDate <= effectiveEndDate
  );

  const rows = [];
  let cdiFactor = 1.0;
  for (const currentDate of timeline) {
    const effectiveQuote = quoteResolver.lookup(currentDate);
    const usdVariation = effectiveQuote.value / initialQuote.value - 1;
    const cdiVariation = cdiFactor - 1;
    const usdPercentVariation = (1 + cdiVariation) / (1 + usdVariation) - 1;

    rows.push({
      date: currentDate,
      cdi_accumulated: cdiVariation * 100,
      usd_accumulated: usdVariation * 100,
      usd_percent_variation: usdPercentVariation * 100,
      adjusted_capital: normalizedInitialBrl * cdiFactor,
      usd_brl_quote: effectiveQuote.value
    });

    if (currentDate < effectiveEndDate) {
      cdiFactor *= 1 + cdiValues.get(currentDate) / 100;
    }
  }
  return rows;
}
